### ISSUE TITLE
Default dataloader type in BaseSGD

### ISSUE THREAD

**Issue by HamedHemati** [ContinualAI/avalanche#1290]
The `BaseSGD` template uses `TaskBalancedDataLoader` by default which is fine for most task incremental settings since there is only one task at a time, but strategies like JointTraining are also based on `BaseSGD`. The main concern here is that it forces `oversample_small_groups=True`, which means if the variation of dataset lengths is high between tasks, the smaller tasks will be repeated many times and can easily overfit. I'm not sure if forcing task-balanced loaders is recommended in joint-training baselines because not all of them do it. I think it makes sense to make it optional and let the user decide whether they want to have a task-balanced or just a normal loader with shuffled samples.  

We can add a new option for data loading type. What do you think?

**Comment by AntonioCarta** [comment]
In general, I think task balancing is a good default in CL. Of course, a default behavior cannot be good in every situation. The current dataloader is already optional because the user can already change the dataloader with subclassing or with a plugin.

Personally, I'm against adding the option in the costructor because dataloaders have many options and we can't fit them all in the constructor.

The disadvantage of the current approach is that the dataloading policy is hidden in the implementation details (same as the optimizer update, model's adaptation and any other implementation detail).
I understand that this is an issue in Avalanche so if you have a better proposal I'm open to discussion, even though I'm not convinced by adding the option in the constructor. Maybe we just need to document better the internals of the strategy or how to override it.

**Comment by HamedHemati** [comment]
True, it's not a big deal, but as you mentioned it needs to be highlighted in the documentation so that the user knows that the default loader is task-balanced and not a regular PyTorch dataloader.  

I was not necessarily suggesting changing the BaseSGD template, but instead, maybe adding the option to the JointTraining class. Because it's the only strategy that is probably affected greatly by choice of dataloader type. 

* This is still a low-priority issue, I just added it to have it on the issue list.

**Comment by AntonioCarta** [comment]
> it's the only strategy that is probably affected greatly by choice of dataloader type.

I think this where we disagree. I think the dataloading policy is an important part of training, both for performance and for generalization. You need to override it often and there are many (too many) possibilities to include them all in the constructor arguments.

That's why I prefer overriding/plugins. You modify the dataloader attribute directly (one thing which gives you total control) instead of passing the arguments (many things which give you limited control). The disadvantage is that overriding and plugins are more complex.

**Comment by HamedHemati** [comment]
Ok, that also makes sense to me. Then we can just need to update the documentation.

**Comment by AntonioCarta** [comment]
Ok, let's focus on the docs. I'm assigning this to you for now.
